### ISSUE TITLE
Spectrogram.iSTFT backward very slow

### ISSUE THREAD

**Issue by ABC0408** [KinWaiCheuk/nnAudio#88]
Setting: Spectrogram.iSTFT(n_fft=1024, win_length=1024, freq_bins=None, hop_length=300,
                                       fmax=7600, fmin=80, sr=24000, trainable_window=False, trainable_kernels=False,
                                       verbose=False)
I used Spectrogram.iSTFT convert spectrograms back to waveforms, most batch backward is faster, but sometimes backward is very slow (0.116s VS 184s). When I used torch.istft instead, backward only took 0.07s.

**Comment by KinWaiCheuk** [comment]
May I know is there anything special about that batch that causes the slowness? I am unable to reproduce to same behaviour. 

What is the GPU you are using? How much memory does it have?

**Comment by ABC0408** [comment]
I'm sure no special operation in the batch, because when I changed to torch.istft, everything is fine.
I'm using one GeForce RTX 3080, total 10015 MB memory.

**Comment by KinWaiCheuk** [comment]
Is it possible for you to write a dummy code, for example creating a random batch of input via `x = torch.randn(16, 44100)`, where 16 is the batch size, or you can change it to the number that will cause the problem; and 44100 is the audio length.
Then you feed this dummy input to the STFT and then iSTFT again, to see if you also face the same problem.

Alternative, if you only use iSTFT, maybe you can create your own dummy spectrogram and feed it to the iSTFT.
If you face the same problem, can you send that code to me? I will analyse it and try to figure out where is the problem.

**Comment by ABC0408** [comment]
Thank you very much.
This problem can not be reproduced. 
I checked my losses and comment some of them, the speed looks normal.

**Comment by ABC0408** [comment]
I still did solve this problem, today I found when GPU-Util reached 100%, loss.backward() got stuck, otherwise it runs normally, this problem happened periodically.

**Comment by KinWaiCheuk** [comment]
It seems to be a GPU bottleneck to me. Can you check when you use torch.istft, what is the GPU utility? Does it reaches 100%?

If torch.istft does not reach 100%, then I think the reason is that nnAudio is faster, but it takes up more GPU resources. So it might hit the GPU bottleneck sometimes.

**Comment by ABC0408** [comment]
I'm using GeForce RTX 2080, the same code everything is fine.

**Comment by KinWaiCheuk** [comment]
> I'm using GeForce RTX 2080, the same code everything is fine.

If this problem only appears when using RTX 3080, but not on RTX 2080, then it seems this problem is somehow related to hardware...

Unfortunately I don't have a RTX 3080 for me to investigate the problem. But thanks for pointing it out, I will bear this in mind.
I am going to reopen this in case other people encounter this issue.

**Comment by ABC0408** [comment]
> It seems to be a GPU bottleneck to me. Can you check when you use torch.istft, what is the GPU utility? Does it reaches 100%?
> 
> If torch.istft does not reach 100%, then I think the reason is that nnAudio is faster, but it takes up more GPU resources. So it might hit the GPU bottleneck sometimes.

Yes, torch.istft doesn't reach 100% on RTX 3080. 
And when I use nnAudio on RTX 2080, the gpu utility always around 95%, not reaching 100%.

**Comment by ABC0408** [comment]
I test a RTX 3090, it face the same issue, @KinWaiCheuk do you have any plans or suggestions to solve?

**Comment by KinWaiCheuk** [comment]
Currently this problem happens in RTX 3090 and RTX 3080, but not on RTX 2080? I am not sure if there is something in the RTX 3000 series that is fundamentally different from the RTX 2000 series.

Have you tried a smaller batch size? (Although we are not facing a memory issue, but it is still worth a try)

**Comment by KinWaiCheuk** [comment]
I did a quick search, and I found that other people are facing the same problem using PyTorch on a RTX 3090. Since nnAudio is implemented using convolutional layers from PyTorch, it faces the same problem. 
https://github.com/pytorch/pytorch/issues/47039

It seems it is related to the CUDA or the cuDNN version. One guy said that he got a speed improvement by upgrading cudnn from 8.0.4 to 8.0.5. Maybe you can give it a try?

**Comment by ABC0408** [comment]
> Currently this problem happens in RTX 3090 and RTX 3080, but not on RTX 2080? I am not sure if there is something in the RTX 3000 series that is fundamentally different from the RTX 2000 series.
> 
> Have you tried a smaller batch size? (Although we are not facing a memory issue, but it is still worth a try)

Yes. Even I tried very small batch size, the speed also not stable, sometimes 0.3 steps/s, somtimes 4 steps/s.

**Comment by ABC0408** [comment]
> I did a quick search, and I found that other people are facing the same problem using PyTorch on a RTX 3090. Since nnAudio is implemented using convolutional layers from PyTorch, it faces the same problem.
> [pytorch/pytorch#47039](https://github.com/pytorch/pytorch/issues/47039)
> 
> It seems it is related to the CUDA or the cuDNN version. One guy said that he got a speed improvement by upgrading cudnn from 8.0.4 to 8.0.5. Maybe you can give it a try?


This is my environment:
torch.__version__:1.8.1,
torch.version.cuda:11.1,
**torch.backends.cudnn.version():8005**

(base) :~$ nvcc -V
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Mon_Nov_30_19:08:53_PST_2020
Cuda compilation tools, release 11.2, V11.2.67
Build cuda_11.2.r11.2/compiler.29373293_0

Python 3.8.5
[GCC 7.3.0] :: Anaconda, Inc. on linux

**Comment by KinWaiCheuk** [comment]
I am afraid that there is nothing we can do at the moment. We might need to wait for NVIDIA to release new version of CUDA or cudnn. I hope they will fix the optimisation issue on RTX 3000 series GPUs... 

The only thing you can do now is to stick with RTX 2080 first...

**Comment by ABC0408** [comment]
> I am afraid that there is nothing we can do at the moment. We might need to wait for NVIDIA to release new version of CUDA or cudnn. I hope they will fix the optimisation issue on RTX 3000 series GPUs...
> 
> The only thing you can do now is to stick with RTX 2080 first...

Okay, thank you very much.

**Comment by turian** [comment]
@ABC0408 Just curious if you post a colab, to see what the performance differences are there

**Comment by ABC0408** [comment]
> @ABC0408 Just curious if you post a colab, to see what the performance differences are there
I tried to use a small model and a random batch inputs to reproduce it, but it didn't appear.
Or some part of my model as a condition, triggering this problem.
Later maybe I need to slowly simplify the network to find this "condition".

**Comment by ABC0408** [comment]
Today I found when I use torch.cuda.amp.autocast() GPU utility reaches 100%, otherwise the speed is normal.
